### PR TITLE
#5: PoC for uasyncio based app switching

### PR DESCRIPTION
**Pull request by MatthewWilkes**
@tomsci What do you think of this app API?

### CODE REVIEW

**Comment by MatthewWilkes** [comment]
@tomsci Added app.py

Yeah, the boot menu description makes sense to me, maybe treat this as a view on how the graphical launcher would work, and we can keep a simple boot menu for emergencies.

**Comment by MatthewWilkes** [comment]
@tomsci I've extended the current apps to follow this pattern, created a distinction between boot menu and a new launcher, and added in switching out of the task. I also changed many things to class-based code rather than module globals, including buttons to reduce cross-app impact.

I know this is more complex an app interface than import-time side-effects or a single `main()` function, but I really like the idea of simplifying the interface especially around wake/sleep, given the desire to drop to sleep mode a lot.

**Comment by MatthewWilkes** [comment]
I've unmarked this as draft, I'd be happy for this to be merged if others agree.

File: modules/app.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        """Await this to pause execution until the specified app
+        is active. Returns a boolean which is True if the app was active
+        or False if we waited for it to become active. """
+        # Infinite loop while the app is inactive, but with
+        # a sleep in the middle so it doesn't use resources.
+        was_already_active = self._current_app == app
+        while self._current_app != app:
+            task = self._awaitables.get(
+                app,
+                uasyncio.create_task(self.sleep())
```
Comment (tomsci) [inline modules/app.py:24]: I don't follow the purpose of the 5 minute sleep task here?

File: modules/app_launcher/__init__.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        self.on_stop()
+    
+    def launch(self, module, app):
+        self.to_launch = module, app
+
+    def on_wake(self):
+        self.cls()
+
+    def update(self):
+        if tidal.JOY_DOWN.value() == 0:
```
Comment (tomsci) [inline modules/app_launcher/__init__.py:58]: should we update this to use Buttons.on_press too?

File: modules/tidal.py [hunk 3]
```diff
@@ -14,6 +14,7 @@
 
 BUTTON_A = Pin(2, Pin.IN, Pin.PULL_UP)
 BUTTON_B = Pin(1, Pin.IN, Pin.PULL_UP)
+BUTTON_FRONT = Pin(6, Pin.IN, Pin.PULL_UP)
```
Comment (tomsci) [inline modules/tidal.py:17]: dammit now I have to go wire up another button on my breadboard :)

**Review by tomsci — COMMENTED** [review]
I like all the object-y refactors and, in principle, the uasyncio stuff too. I think we need to think a bit more about the `interval` loop and maybe see if we can get something more fully event driven.

Also we may need to think more about app lifetimes. There doesn't seem to be a way of shutting apps down atm (short of rebooting the board) and we might want to make that the default when BUTTON_FRONT is pressed unless the app declares `self.run_in_background=True` or something, and push that logic into App or Launcher rather than having the check for BUTTON_FRONT be something each app's `update()` fn has to do itself.

That said I'm ok with merging this now if you want to get this landed sooner rather than later.

**Pull request merged**